docs/proposals/20220125-ipam-integration.md

```diff
@@ -0,0 +1,389 @@
+
+When an IPAM provider finds a IPAddressClaim that references one of its Pools, it allocates an IP address, creates an IPAddress resource and updates the status of the IPAddressClaim with a reference to this IPAddress.
+
+#### Examples
+
+A simple example for an IP Pool could look like this:
+
+```yaml
+apiVersion: ipam.cluster.x-k8s.io/v1alpha1
+kind: InClusterIPPool
+metadata:
+  name: some-pool
+spec:
+  pools:
+    - subnet: 10.10.10.0/24
+      start: 10.10.10.100
+      end: 10.10.10.200
+```
+A pool for an Infoblox Provider could look like this:
+
+```yaml
+apiVersion: ipam.cluster.x-k8s.io/v1alpha1
+kind: InfobloxIPPool
+metadata:
+  name: ib-pool
+spec:
+  networkView: "some-view"
+  dnsZone: "test.example.com."
+  network: 10.10.10.0/24
+```
+
+### Consumption
+
+The consumers of this API contract are the infrastructure providers. As mentioned in the introduction, not all providers will require integration with IPAM systems. In addition, DHCP may be a viable or even better option in some environments. Integration with this API should therefore be implemented as an optional feature.
+
+To support IPAM integration, two things need to be implemented:
+
+1. Users need to be able to specify the IP Pool to allocate addresses from.
+2. The infrastructure provider needs to create IPAddressClaims for all ip addresses it requires, and then needs to wait until they are fulfilled.
+
+The relationships between the provider and IPAM resources are shown in the diagram below.
+
+![relationships of ipam custom resources](images/ipam-integration/consumption.png)
+
+More implementation details can be found [here](#consuming-as-an-infrastructure-provider).
+
+#### Example
+
+Using CAPV as an example, a template *could* look like this:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: VSphereMachineTemplate
+metadata:
+ name: example
+ namespace: vsphere-site1
+spec:
+ template:
+   spec:
+     cloneMode: FullClone
+     numCPUs: 8
+     memoryMiB: 8192
+     diskGiB: 45
+     network:
+       devices:
+       - dhcp4: false
+         fromPool: # reference to the pool
+           group: ipam.cluster.x-k8s.io/v1alpha1
+           kind: IPPool
+           name: testpool
+```
+
+A Machine generated from that template could then look like this:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: VSphereMachine
+metadata:
+name: example-1
+namespace: vsphere-site1
+spec:
+ cloneMode: FullClone
+ numCPUs: 8
+ memoryMiB: 8192
+ diskGiB: 45
+ network:
+   devices:
+   - dhcp4: false
+     fromPool: # reference to the pool
+       group: ipam.cluster.x-k8s.io/v1alpha1
+       kind: IPPool
+       name: testpool
+status:
+ network:
+   devices:
+     - claim:
+         name: example-1-1
+```
+
+### User Stories
```

Comment:
Should we move this section to the beginning of the Proposal section?

Reply:
I've moved it up for now.